### ISSUE TITLE
Portal to request extra permissions

### ISSUE THREAD

**Issue by fenryxo** [flatpak/xdg-desktop-portal#205]
It would be great if an app could specify minimal permissions in metadata and then ask for extra when necessary.

This would be especially useful for features that break backward compatibility, e.g. Bluetooth (flatpak/flatpak#1719). E.g., my app can optionally use bluetooth but I cannot just add `--allow=bluetooth` because then the app would fail to run with older flatpak versions (flatpak/flatpak#1931).

**Comment by matthiasclasen** [comment]
what you describe is basically what portals are: a way to request extra access at runtime.

**Comment by fenryxo** [comment]
I updated the description to be clearer. I mean permissions in the sense of `--allow=bluetooth` or `--share=network` rather than "take a screenshot" or "open a file".

**Comment by matthiasclasen** [comment]
Those can't be made dynamic, because they have to be taken into account when the sandbox is set up.

**Comment by matthiasclasen** [comment]
But you raise a good point wrt to backwards compatibility. We need to take a serious look at that.

**Comment by fenryxo** [comment]
> Those can't be made dynamic, because they have to be taken into account when the sandbox is set up.

Then the new permission could be stored and a user could be informed that it will be applied after the app is restarted.

**Comment by matthiasclasen** [comment]
So, alex told me that we *are* ignoring unknown permissions, but only since 0.10.something.

**Comment by fenryxo** [comment]
>So, alex told me that we are ignoring unknown permissions, but only since 0.10.something.

Indeed, since 0.10.3. That's good, but 0.10.3 is too new. E.g., Debian stable has 0.8.9 and I'm reluctant to burden my users with enabling Debian backports. I know that the package is perfectly maintained but it is still an extra step for an optional feature, an extra step in the to-install-or-not-to-install decision process.

So, when I add Bluetooth functionality to my app, I'll need to ask a user to enable Bluetooth via `flatpak override`, which is bad user experience. I'd appreciate if there was a Flatpak dialog for that (with better wording): 

"Blah blah Application asks for Bluetooth permission. It will be stored permanently, but you can disable it anytime." [Grant Bluetooth access] [Deny] → "The application needs to be restarted to apply new settings."

Another benefit would be the context. When the app is installed with mandatory Bluetooth, a user might wonder why the app needs Bluetooth. When the app is installed without Bluetooth and the user is asked to grant Bluetooth permission after enabling "Remote control over Bluetooth", there is no surprise. I think this would be much better user experience.

**Comment by alexlarsson** [comment]
I realize 0.8 is out there (its in rhel too), but rather than adding stuff to the API that we have to support forever (even though its useless post 0.10.3) we should get 1.0 out and in use in as many places as possible.

**Comment by fenryxo** [comment]
> [...]  adding stuff to the API that we have to support forever (even though its useless post 0.10.3) [...]

I don't think the proposed API is entirely useless post 0.10.3. The ability to ask for extra permissions in a better context (like Android does nowadays), as I described in my previous post, is still a valid use case post 0.10.3. The question is then whether the feature is worth adding & maintaining or not.

> [...] we should get 1.0 out and in use in as many places as possible.

That would be great :-)

**Comment by hadess** [comment]
I don't think that the application needs to be able to *ask* for more permissions, but it needs to be able to detect whether some permissions it is packaged/deployed with (say, network access, or Bluetooth access) have been removed.

The pattern of directing users towards the settings for missing permissions is something that's well established in mobile OSes, but only for the equivalent of "portal" permissions. For the case of Bluetooth applications, the application would need to be able to detect the "bluetooth" access from the `finish-args` being missing, any additional permissions should be obtained via a Bluetooth portal (which doesn't exist yet).

The less `finish-args` type permissions we use, the better it is for future-proofing device and feature access outside the sandbox.

**Comment by matthiasclasen** [comment]
I believe apps can read this information from /.flatpak-info, inside the sandbox

**Comment by indolering** [comment]
Did this ever get implemented?  

> what you describe is basically what portals are: a way to request extra access at runtime.

There are a large number of applications which don't have native support for portals, so you either end up with a "sandbox" that includes R/W access to the [entire `/home` directory](https://github.com/flathub/org.signal.Signal/issues/20) or [users complaining about not being able to access their files](https://github.com/flathub/im.riot.Riot/issues/33).  Yes, upstream software *should* switch to using native portals, but they don't give a shit about security.  For example, the [ticket](https://github.com/electron/electron/issues/2911) to add this functionality to Electron is 5 years old.  There was [a PR made in 2018](https://github.com/electron/electron/pull/15293) which stalled out and a second [PR](https://github.com/electron/electron/pull/19159) was first posted nearly a year ago.

This feature would make it easier for app developers to request access without having to patch the upstream frameworks they are using.

**Comment by TingPing** [comment]
@indolering This issue never made sense to implement and doesn't help in the examples you give.

**Comment by indolering** [comment]
> Please don't bump ancient bugs.

It was closed without an explanation, I was asking for clarification.
